core/contracts/Store.sol

```diff
@@ -0,0 +1,121 @@
+/*
+  Store implementation.
+ 
+  An implementation of StoreInterface with a fee per second and withdraw functions for the owner.
+*/
+pragma solidity ^0.5.0;
+pragma experimental ABIEncoderV2;
+
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import "./StoreInterface.sol";
+import "./Withdrawable.sol";
+import "./FixedPoint.sol";
+import "./MultiRole.sol";
+
+
+/** 
+ * @title An implementation of StoreInterface that can accept Oracle fees in ETH or any arbitrary ERC20 token.
+ */
+contract Store is StoreInterface, MultiRole {
+
+    using SafeMath for uint;
+    using FixedPoint for FixedPoint.Unsigned;
+    using FixedPoint for uint;
+
+    enum Roles {
+        Governance
+    }
+
+    FixedPoint.Unsigned private fixedOracleFeePerSecond; // Percentage of 1 E.g., .1 is 10% Oracle fee.
+
+    uint private weeklyDelayFee;
+    mapping(address => FixedPoint.Unsigned) private finalFees;
+    uint private constant SECONDS_PER_WEEK = 604800;
+
+    // TODO(roz): Used to make doubly sure that roles are initialized only once. 
+    // Figure out what's going wrong with coverage to necessitate this hack.
+    bool private rolesInitialized;
+
+    constructor() public {
+        initializeRolesOnce();
+    }
+
+    function payOracleFees() external payable {
+        require(msg.value > 0);
+    }
+
+    function payOracleFeesErc20(address erc20Address) external {
+        IERC20 erc20 = IERC20(erc20Address);
+        uint authorizedAmount = erc20.allowance(msg.sender, address(this));
+        require(authorizedAmount > 0);
+        require(erc20.transferFrom(msg.sender, address(this), authorizedAmount));
+    }
+
+    /**
+     * @dev Sets a new weekly delay fee
+     */ 
+    function setWeeklyDelayFee(uint newWeeklyDelayFee) 
+        external 
+        onlyRoleHolder(uint(Roles.Governance)) 
+    {
+        weeklyDelayFee = newWeeklyDelayFee;
+    }
+
+    function computeRegularFee(uint startTime, uint endTime, FixedPoint.Unsigned calldata pfc) 
+        external 
+        view 
+        returns (FixedPoint.Unsigned memory regularFee, FixedPoint.Unsigned memory latePenalty) 
+    {
+        uint timeDiff = endTime.sub(startTime);
+
+        regularFee = pfc.mul(fixedOracleFeePerSecond).mul(timeDiff);
+        latePenalty = pfc.mul(weeklyDelayFee.mul(timeDiff.div(SECONDS_PER_WEEK)));
+
+        return (regularFee, latePenalty);
+    }
+
+    function computeFinalFee(address currency) 
+        external 
+        view 
+        returns (FixedPoint.Unsigned memory finalFee) 
+    {
+        finalFee = finalFees[currency];
+    }
+
+    /**
+     * @dev Sets a new oracle fee per second
+     */ 
+    function setFixedOracleFeePerSecond(FixedPoint.Unsigned memory newOracleFee) 
+        public 
```

Comment:
nit: why not `external`?

Reply:
making this function external would mean the param would need to be calldata not memory which upsets the compiler

Reply:
Oh right - good point.